main/lsp/ErrorReporter.cc

```diff
@@ -0,0 +1,104 @@
+#include "main/lsp/ErrorReporter.h"
+#include "core/lsp/TypecheckEpochManager.h"
+#include "main/lsp/LSPMessage.h"
+#include "main/lsp/LSPOutput.h"
+#include "main/lsp/json_types.h"
+
+namespace sorbet::realmain::lsp {
+using namespace std;
+ErrorReporter::ErrorReporter(shared_ptr<const LSPConfiguration> config) : config(move(config)) {}
+
+// Used for unit tests
+const std::vector<ErrorStatus> &ErrorReporter::getFileErrorStatuses() const {
+    return fileErrorStatuses;
+};
+
+vector<core::FileRef> ErrorReporter::filesUpdatedSince(u4 epoch) {
+    vector<core::FileRef> filesUpdatedSince;
+    for (size_t i = 1; i < fileErrorStatuses.size(); ++i) {
+        ErrorStatus fileErrorStatus = fileErrorStatuses[i];
+        if (fileErrorStatus.lastReportedEpoch >= epoch && fileErrorStatus.hasErrors) {
+            filesUpdatedSince.push_back(core::FileRef(i));
+        }
+    }
+    return filesUpdatedSince;
+}
+
+void ErrorReporter::pushDiagnostics(u4 epoch, core::FileRef file, const vector<unique_ptr<core::Error>> &errors,
+                                    const core::GlobalState &gs) {
+    ENFORCE(file.exists());
+    if (file.data(gs).epoch > epoch) {
```

Comment:
Invariant: If a file's `epoch` is updated, then we will pushDiagnostics for it in that epoch.

Therefore, if `file.data(gs).epoch > epoch`, then `fileErrorStatus.lastReportedEpoch > epoch`.

So, we can safely remove this check.